### PR TITLE
Warn if spaces instead of tabs are detected to prevent confusion

### PR DESCRIPTION
**Pull request by pd3**
… as in https://github.com/samtools/bcftools/issues/1575

### CODE REVIEW

File: vcf.c [hunk 1]
```diff
@@ -145,12 +145,34 @@ int bcf_hdr_add_sample(bcf_hdr_t *h, const char *s)
...
+            if ( i==0 && strncmp(p,"#CHROM",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==1 && strncmp(p,"POS",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==2 && strncmp(p,"ID",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==3 && strncmp(p,"REF",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==4 && strncmp(p,"ALT",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==5 && strncmp(p,"QUAL",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==6 && strncmp(p,"FILTER",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==7 && strncmp(p,"INFO",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( i==8 && strncmp(p,"FORMAT",q-p) ) hts_log_error("Could not parse the sample line: %s",str);
+            if ( !hdr_space_warned ) hts_log_warning("The sample names contain spaces, is this intended? Trying to proceed anyway..");
```
Comment (jkbonfield) [inline vcf.c:165]: I assume you added ".." so it the function can add the final ".", but it triggers a failure the test harness.  Maybe just omit the ".." to silence this.

Comment (pd3) [reply vcf.c]: The dots ".." are removed now

**Comment by pd3** [comment]
Thanks for your HO.

HTSlib and BCFtools often have to work with malformed VCFs. The question is, should HTSlib attempt to do some error recovery and proceed anyway, or be strict. In this case my opinion is that it should try to proceed. Note that your test case ignores the fact that the current develop version accepts everything as long as `#CHROM\tPOS` is correct and does not give the user any clue what may be the problem.

I agree that the PR version should be improved so that the warning message makes a distinction between sample names and the standard VCF columns, or at least mentions both.

**Comment by jmarshall** [comment]
My test case and comment were a review of the proposed PR, not of the status quo on **develop**. As you say, there is room for improvement on the status quo.

**Comment by jkbonfield** [comment]
Given it's malformed data and the PR is already clearly a significant improvement on develop,  I think the minimal change of making the diagnostic warning message more encompassing is a simple fix.  (Ie option 2) Eg mentioning spaces on `#CHROM` line without pinning it down specifically to being within the samples component.  It's stil sufficient for the user to identify the problem.

FWIW I'm not against the idea of fixing the error while emitting a warning.  I don't view this as incompatible with htsjdk as what's then output is compatible with all tools once again (assuming nothing else is broken).  I'm rather ambivalent as to whether it's a good thing to fix-and-warn or hard-error though.  I guess it depends on the prevalence of historic data out there having this defect in which case fix is good, or whether it's some recent tool with buggy output in which case error is good and file a bug report.

**Comment by jmarshall** [comment]
> I don't view this as incompatible with htsjdk as what's then output is compatible with all tools once again (assuming nothing else is broken).

I perhaps shouldn't have mentioned the output, as that's a red herring.

The point is that if people archive files like the example input file on the basis that they work with htslib/bcftools, they or others will later find out that they don't work with the Java tools. This would be an interoperability problem caused by htslib accepting such invalid files.

**Comment by pd3** [comment]
I just modified the code to strictly conform to VCF specification + print an informative error if not, as was discussed yesterday at the meeting.

**Comment by jkbonfield** [comment]
I note it doesn't accept "FORMAT\n", so with zero samples.  I'm assuming this is correct as the specification says "_If genotype data is present in the file, these are followed by a FORMAT column header, then an arbitrary number
of sample IDs_".  While "arbitrary number" doesn't exclude zero, and it says "If" and not "If and only if", I think it's pretty clear that's the intent.

**Comment by pd3** [comment]
I had given it a thought while reviewing the code and on several occasions before. Practically speaking, FORMAT without samples is useless. Formally, the VCF specification says "If genotype data is present in the file, these are followed...". This implies that when there are no genotype data (zero samples), nothing follows the INFO field. Suggesting otherwise is nitpicking.

**Comment by jkbonfield** [comment]
I just noted it incase it was something to consider as the spec doesn't explicitly forbid this case, but I'm glad it's already been considered and we agree too on it not being a problem.

**Pull request merged**